pkg/vsphere/rbac/rbac.go

```diff
@@ -163,6 +169,65 @@ func (am *AuthzManager) PrincipalBelongsToGroup(ctx context.Context, group strin
 	return false, nil
 }
 
+// PrincipalHasReadOnlyPrivs returns true if the user (principal) in the AuthzManager has
+// at least a read-only role on the input ref object, false otherwise.
+func (am *AuthzManager) PrincipalHasReadOnlyPrivs(ctx context.Context, ref types.ManagedObjectReference) (bool, error) {
+	perms, err := am.GetPermissions(ctx, ref)
+	if err != nil {
+		return false, err
+	}
+
+	roles, err := am.RoleList(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	opsUser := strings.ToLower(am.Principal)
+
+	// privMap keeps track of privileges held by the opsUser on the input ref object.
+	privMap := make(map[string]struct{})
+
+	// These vars keep track of whether the corresponding privileges have been found.
+	var anon, read, view bool
+
+	for _, perm := range perms {
+		// Skip permission entries assigned to other users.
+		user := am.formatPrincipal(perm.Principal)
+		if user != opsUser {
+			continue
+		}
+
+		role := roles.ById(perm.RoleId)
+		if role == nil {
+			return false, fmt.Errorf("unable to find role %d by ID", perm.RoleId)
+		}
+
+		// Check the role for privileges that satisfy the read-only role.
+		privs := role.Privilege
+		for i := range privs {
```

Comment:
Do you really need this map?  It can be done a lot simpler.
```Go
for _, v := range privs {
   anon = anon || v == sysAnonPriv
   read = read|| v == sysReadPriv
   view = view|| v == sysViewPriv
}
....
return anon && read && view, nil
```